solo-io/gloo
# Ensure invalid virtual hosts are not duplicated on proxy

# Description
- Fixed logic in our proxy reconciler, which allowed invalid virtual hosts to be added to a proxy twice. 
- Add some utility methods for our Kube2e tests, to make it easier for developers to add logic around waiting for resources to be accepted and cleaned up
- Fix test pollution that did lead to flakes
- Fix robustness_test, which was written in a way that relied on the exact bug that this PR fixes, to be present
- Fix disable_transformation_test (https://github.com/solo-io/gloo/issues/4900) which would flake occasionally, and did during the development of this PR 

# Context

### Mix of valid and invalid virtual services
Example logs: https://console.cloud.google.com/cloud-build/builds/a13c32e3-fddf-4263-b28a-e931933d9f87;step=14?project=solo-public

This test flake was due to a bug in our proxy_reconciler. We incorrectly were duplicating virtual hosts on the proxy object. The proxy only transitioned in certain situations (depending on metadata of the object) which is why this didn't occur more frequently. However, when it did happen, it caused this flaking test.

I added a unit test to verify the change that I made fixes the problem. 
I also added a comment to the code so that future developers would understand the reason for the change.

### Does not equal length of map
Sometimes, I ran into a Gateway test (I don't have the logs handy) that produced the following error:
```
[
gloo-system-gateway-443_gloo-system-16567209401833135566:true 
gloo-system-gateway-proxy-443_gloo-system-7590740161429284937:true 
gloo-system-gateway-proxy-80_gloo-system-7992917100079133684:true 
gloo-system-gloo-9976_gloo-system-12285151928708083525:true 
gloo-system-gloo-9977_gloo-system-7484856654061583909:true 
gloo-system-gloo-9979_gloo-system-5451814594336066521:true 
gloo-system-gloo-9988_gloo-system-1559281721111094449:true 
gloo-system-testrunner-1234_gloo-system-10782942849661275980:true 
kube-svc:gloo-system-gateway-443_gloo-system-10160320302259603858:true 
kube-svc:gloo-system-gateway-proxy-443_gloo-system-13361889343733600589:true 
kube-svc:gloo-system-gateway-proxy-80_gloo-system-15324810366434219216:true 
kube-svc:gloo-system-gloo-9976_gloo-system-6098176559075073065:true 
kube-svc:gloo-system-gloo-9977_gloo-system-8667782289304402933:true
 kube-svc:gloo-system-gloo-9979_gloo-system-1598973103534690965:true
 kube-svc:gloo-system-gloo-9988_gloo-system-6721171892144873845:true 
kube-svc:gloo-system-testrunner-1234_gloo-system-1741026987556571732:true
] 

does not equal length of map

[
gloo-system-gateway-443_gloo-system-16567209401833135566:0xc0008e6ce0 
gloo-system-gateway-proxy-443_gloo-system-7590740161429284937:0xc0008e6da0 
gloo-system-gateway-proxy-80_gloo-system-7992917100079133684:0xc0008e6e50 
gloo-system-gloo-9976_gloo-system-12285151928708083525:0xc0008e6ec0 
gloo-system-gloo-9977_gloo-system-7484856654061583909:0xc0008e6f30 
gloo-system-gloo-9979_gloo-system-5451814594336066521:0xc0008e6fa0
gloo-system-gloo-9988_gloo-system-1559281721111094449:0xc0008e7010 
gloo-system-redbluezqw6b-5678_gloo-system-6470157035427515329:0xc0008e74b0
gloo-system-testrunner-1234_gloo-system-10782942849661275980:0xc0008e7060 
kube-svc:gloo-system-gateway-443_gloo-system-10160320302259603858:0xc0008e70d0 
kube-svc:gloo-system-gateway-proxy-443_gloo-system-13361889343733600589:0xc0008e7190 
kube-svc:gloo-system-gateway-proxy-80_gloo-system-15324810366434219216:0xc0008e7270 
kube-svc:gloo-system-gloo-9976_gloo-system-6098176559075073065:0xc0008e72f0 
kube-svc:gloo-system-gloo-9977_gloo-system-8667782289304402933:0xc0008e7360 
kube-svc:gloo-system-gloo-9979_gloo-system-1598973103534690965:0xc0008e73d0 
kube-svc:gloo-system-gloo-9988_gloo-system-6721171892144873845:0xc0008e7440 
kube-svc:gloo-system-redbluezqw6b-5678_gloo-system-5865765942139567028:0xc0008e74d0 
kube-svc:gloo-system-testrunner-1234_gloo-system-1741026987556571732:0xc0008e7490
]
```
I believe this is just due to test pollution, since the mismatched service (ie the service in one map but not the other) is `gloo-system-redbluezqw6b-5678_gloo-system-6470157035427515329:0xc0008e74b0`. There is only one test that uses this service, so I added an Eventually to ensure it is cleaned up after the test 

### Gateway robustness test changes
This test intentionally wrote a virtual service with duplicate domains, which triggers the exact bug that this PR fixes. After introducing the fix, and not changing the test at all, the test consistently fails, because it expects the proxy to enter a Rejected state. This would happen before, but with the fix, invalid virtual hosts are stripped properly, and not duplicated, so the proxy no longer enters a rejected state.

I updated the test, and simplified some of the logic to make it consistent with another test that is validating similar behavior.

### Disable transformation validation kube2e test flake

This flake was due to behavior of the gateway validation server. When the gateway validation settings change (as we do in this test), the server restarts and we need to afford it enough time to reboot. I re-used a pattern we have relied on in other tests, to ensure that the settings are given enough time to propagate before moving forward with the test execution.

# Checklist:


# Testing
I have another branch with the tests re-written in a way to make running a single test over and over easier/faster. As of now, we can supply the `untilItFails` flag to ginkgo, but since Kube2e tests install gloo and uninstall it before/after Suites, I re-wrote the tests so that I could repeat tests without having to re-install gloo each time. This was just useful for debugging, and I didn't add the code to the PR. If we're interested in a true debugging utility for kube2e tests, there are likely better ways to set that up. The branch is: https://github.com/solo-io/gloo/tree/regression-test-flakes

Here are the steps you can run to run these tests quickly:
1. Checkout code from testing branch: `git checkout regression-test-flakes`
1. Checkout code from this branch: `git checkout fix-regression-test-flakes`
1.  Build images with version, and push to kind cluster. `VERSION=v1.8.2 CLUSTER_NAME=regression-flakes CLUSTER_NODE_VERSION=v1.21.1 ./ci/deploy-to-kind-cluster.sh`
1. Checkout files from test branch: `git checkout regression-test-flakes test/kube2e/util.go`,  `git checkout regression-test-flakes test/kube2e/gateway/gateway_test.go`
1. Run focused regression tests: ` KUBE2E_TESTS=gateway TEAR_DOWN=true WAIT_ON_FAIL=1 make run-ci-regression-tests
`

This will run the previously flaking test 100 times in a row. Before this change (you can verify for yourself), if I ran this, most times a flake would occur within the 100 times. However, with this change, I consistently complete all 100 without an error.
